File: apps/patter-voice-sdk/libraries/typescript/src/providers/speechmatics-stt.ts
/**
 * Speechmatics Speech-to-Text adapter for the Patter SDK pipeline mode.
 *
 * Streams PCM audio to the Speechmatics real-time WebSocket API
 * (`wss://eu.rt.speechmatics.com/v2`) and emits Patter-normalised
 * {@link Transcript} events. Mirrors `SpeechmaticsSTT` in the Python SDK.
 *
 * Divergence from Python: the Python adapter wraps the official
 * `speechmatics-voice` Python SDK (Voice Agent presets, smart turn
 * detection, etc.). No equivalent Node SDK is published, so this TypeScript
 * adapter speaks the underlying RT v2 wire protocol directly via `ws`.
 * The user-facing options (`turnDetectionMode`, `endOfUtteranceSilenceTrigger`,
 * `maxDelay`, `enablePartials`, `enableDiarization`, `additionalVocab`,
 * `operatingPoint`, `domain`, `outputLocale`, `language`, `sampleRate`) all
 * map 1:1 onto the Python adapter so callers can switch SDKs without
 * changing their config.
 */

import WebSocket from 'ws';
import type { IncomingMessage } from 'http';
import { AuthenticationError, PatterConnectionError, RateLimitError } from '../errors';
import { getLogger } from '../logger';

/** Patter-normalised transcript event emitted by {@link SpeechmaticsSTT}. */
export interface Transcript {
  readonly text: string;
  readonly isFinal: boolean;
  readonly confidence: number;
}

type TranscriptCallback = (transcript: Transcript) => void;
type ErrorCallback = (error: Error) => void;

const SPEECHMATICS_RT_URL = 'wss://eu.rt.speechmatics.com/v2';
const CONNECT_TIMEOUT_MS = 10000;

/**
 * Endpoint / turn-detection handling mode. Mirrors the values accepted by
 * Python's `TurnDetectionMode`. Maps onto Speechmatics's
 * `conversation_config` knobs on the wire.
 */
export const TurnDetectionMode = {
  EXTERNAL: 'external',
  FIXED: 'fixed',
  ADAPTIVE: 'adaptive',
  SMART_TURN: 'smart_turn',
} as const;
export type TurnDetectionMode =
  (typeof TurnDetectionMode)[keyof typeof TurnDetectionMode];

/** Common PCM sample rates for Speechmatics streaming input. */
export const SpeechmaticsSampleRate = {
  HZ_8000: 8000,
  HZ_16000: 16000,
  HZ_44100: 44100,
} as const;
export type SpeechmaticsSampleRate =
  (typeof SpeechmaticsSampleRate)[keyof typeof SpeechmaticsSampleRate];

/** Audio encodings accepted by Speechmatics's real-time API. */
export const SpeechmaticsAudioEncoding = {
  PCM_S16LE: 'pcm_s16le',
} as const;
export type SpeechmaticsAudioEncoding =
  (typeof SpeechmaticsAudioEncoding)[keyof typeof SpeechmaticsAudioEncoding];

/** Speechmatics operating points (accuracy vs latency trade-off). */
export const SpeechmaticsOperatingPoint = {
  ENHANCED: 'enhanced',
  STANDARD: 'standard',
} as const;
export type SpeechmaticsOperatingPoint =
  (typeof SpeechmaticsOperatingPoint)[keyof typeof SpeechmaticsOperatingPoint];

/** Speechmatics RT server-message type names emitted to the client. */
export const SpeechmaticsServerMessage = {
  RECOGNITION_STARTED: 'RecognitionStarted',
  ADD_PARTIAL_TRANSCRIPT: 'AddPartialTranscript',
  ADD_TRANSCRIPT: 'AddTranscript',
  END_OF_UTTERANCE: 'EndOfUtterance',
  END_OF_TRANSCRIPT: 'EndOfTranscript',
  AUDIO_ADDED: 'AudioAdded',
  INFO: 'Info',
  WARNING: 'Warning',
  ERROR: 'Error',
} as const;
export type SpeechmaticsServerMessage =
  (typeof SpeechmaticsServerMessage)[keyof typeof SpeechmaticsServerMessage];

/** Constructor options for {@link SpeechmaticsSTT}. */
export interface SpeechmaticsSTTOptions {
  /** Override the realtime endpoint (default `wss://eu.rt.speechmatics.com/v2`). */
  readonly baseUrl?: string;
  /** BCP-47 language code. Default `"en"`. */
  readonly language?: string;
  /** Endpoint / turn-detection mode. Default `"adaptive"`. */
  readonly turnDetectionMode?: TurnDetectionMode;
  /** PCM sample rate (Hz). Default 16000. */
  readonly sampleRate?: SpeechmaticsSampleRate | number;
  /** Attach speaker IDs to transcripts. Default `false`. */
  readonly enableDiarization?: boolean;
  /** Max latency in seconds before the engine emits finals. Range 0.7..4.0. */
  readonly maxDelay?: number;
  /** Silence (s) that triggers EOU. Range (0, 2). */
  readonly endOfUtteranceSilenceTrigger?: number;
  /** Max EOU delay (s); must exceed `endOfUtteranceSilenceTrigger`. */
  readonly endOfUtteranceMaxDelay?: number;
  /** Include partial transcripts in interim output. Default `true`. */
  readonly includePartials?: boolean;
  /** Additional vocabulary entries (`{content, sounds_like?}`). */
  readonly additionalVocab?: ReadonlyArray<Record<string, unknown>>;
  /** Operating point (`enhanced` | `standard`). */
  readonly operatingPoint?: SpeechmaticsOperatingPoint;
  /** Optional Speechmatics domain (e.g. `"finance"`). */
  readonly domain?: string;
  /** Optional output locale (e.g. `"en-GB"`). */
  readonly outputLocale?: string;
}

interface SpeechmaticsAlternative {
  content?: string;
  confidence?: number;
}

interface SpeechmaticsResult {
  alternatives?: SpeechmaticsAlternative[];
  type?: string;
}

interface SpeechmaticsTranscriptMessage {
  message?: string;
  metadata?: { transcript?: string };
  results?: SpeechmaticsResult[];
  reason?: string;
  type?: string;
}

/**
 * Streaming STT adapter for Speechmatics's RT v2 WebSocket API.
 *
 * @example
 * ```ts
 * const stt = new SpeechmaticsSTT('sm_api_key', { language: 'en' });
 * stt.onTranscript((t) => console.log(t.text, t.isFinal));
 * await stt.connect();
 * stt.sendAudio(pcm16Chunk);
 * stt.close();
 * ```
 */
export class SpeechmaticsSTT {
  /** Stable pricing/dashboard key — read by stream-handler/metrics. */
  static readonly providerKey = 'speechmatics';
  private ws: WebSocket | null = null;
  private readonly transcriptCallbacks = new Set<TranscriptCallback>();
  private readonly errorCallbacks = new Set<ErrorCallback>();
  private running = false;
  /** Sequence number of the last audio chunk acknowledged via `AudioAdded`. */
  private lastSeqNo = 0;

  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly language: string;
  private readonly turnDetectionMode: TurnDetectionMode;
  private readonly sampleRate: number;
  private readonly enableDiarization: boolean;
  private readonly maxDelay: number | undefined;
  private readonly endOfUtteranceSilenceTrigger: number | undefined;
  private readonly endOfUtteranceMaxDelay: number | undefined;
  private readonly includePartials: boolean;
  private readonly additionalVocab: ReadonlyArray<Record<string, unknown>>;
  private readonly operatingPoint: SpeechmaticsOperatingPoint | undefined;
  private readonly domain: string | undefined;
  private readonly outputLocale: string | undefined;

  /** Construction args replayed by clone(). */
  private readonly patterCtorArgs: unknown[];

  constructor(apiKey: string, options: SpeechmaticsSTTOptions = {}) {
    this.patterCtorArgs = [apiKey, options];
    if (!apiKey) {
      throw new Error('Speechmatics apiKey is required');
    }

    // Validate ranges per the Speechmatics Voice SDK contract — kept in
    // lockstep with the Python adapter so misconfigurations surface the
    // same way regardless of which SDK the user picks.
    const eouSilence = options.endOfUtteranceSilenceTrigger;
    const eouMax = options.endOfUtteranceMaxDelay;
    const maxDelay = options.maxDelay;
    if (eouSilence !== undefined && !(eouSilence > 0 && eouSilence < 2)) {
      throw new Error('endOfUtteranceSilenceTrigger must be between 0 and 2');
    }
    if (
      eouMax !== undefined &&
      eouSilence !== undefined &&
      eouMax <= eouSilence
    ) {
      throw new Error(
        'endOfUtteranceMaxDelay must be greater than endOfUtteranceSilenceTrigger',
      );
    }
    if (maxDelay !== undefined && !(maxDelay >= 0.7 && maxDelay <= 4.0)) {
      throw new Error('maxDelay must be between 0.7 and 4.0');
    }

    this.apiKey = apiKey;
    this.baseUrl = options.baseUrl ?? SPEECHMATICS_RT_URL;
    this.language = options.language ?? 'en';
    this.turnDetectionMode = options.turnDetectionMode ?? TurnDetectionMode.ADAPTIVE;
    this.sampleRate = options.sampleRate ?? SpeechmaticsSampleRate.HZ_16000;
    this.enableDiarization = options.enableDiarization ?? false;
    this.maxDelay = maxDelay;
    this.endOfUtteranceSilenceTrigger = eouSilence;
    this.endOfUtteranceMaxDelay = eouMax;
    this.includePartials = options.includePartials ?? true;
    this.additionalVocab = options.additionalVocab ?? [];
    this.operatingPoint = options.operatingPoint;
    this.domain = options.domain;
    this.outputLocale = options.outputLocale;
  }

  /** Build the JSON `StartRecognition` payload sent on connect. */
  private buildStartRecognition(): Record<string, unknown> {
    const transcriptionConfig: Record<string, unknown> = {
      language: this.language,
      enable_partials: this.includePartials,
      diarization: this.enableDiarization ? 'speaker' : 'none',
    };
    if (this.maxDelay !== undefined) transcriptionConfig.max_delay = this.maxDelay;
    if (this.operatingPoint !== undefined) {
      transcriptionConfig.operating_point = this.operatingPoint;
    }
    if (this.domain !== undefined) transcriptionConfig.domain = this.domain;
    if (this.outputLocale !== undefined) {
      transcriptionConfig.output_locale = this.outputLocale;
    }
    if (this.additionalVocab.length > 0) {
      transcriptionConfig.additional_vocab = [...this.additionalVocab];
    }

    // `conversation_config` carries the turn-detection knobs used by the
    // Python Voice SDK. Speechmatics RT v2 accepts these fields on the
    // root `transcription_config` object.
    const conversationConfig: Record<string, unknown> = {
      end_of_utterance_mode: this.turnDetectionMode,
    };
    if (this.endOfUtteranceSilenceTrigger !== undefined) {
      conversationConfig.end_of_utterance_silence_trigger =
        this.endOfUtteranceSilenceTrigger;
    }
    if (this.endOfUtteranceMaxDelay !== undefined) {
      conversationConfig.end_of_utterance_max_delay = this.endOfUtteranceMaxDelay;
    }
    transcriptionConfig.conversation_config = conversationConfig;

    return {
      message: 'StartRecognition',
      audio_format: {
        type: 'raw',
        encoding: SpeechmaticsAudioEncoding.PCM_S16LE,
        sample_rate: this.sampleRate,
      },
      transcription_config: transcriptionConfig,
    };
  }

  /** Open the streaming WebSocket and send the `StartRecognition` frame. */

  /**
   * Fresh adapter built with this instance's construction arguments —
   * called per call by the stream handler so concurrent calls never share
   * connection state (sockets/queues; cross-call transcript bleed).
   */
  clone(): this {
    const ctor = this.constructor as new (...args: unknown[]) => this;
    return new ctor(...this.patterCtorArgs);
  }

  async connect(): Promise<void> {
    if (this.ws !== null) return;

    const ws = new WebSocket(this.baseUrl, {
      headers: { Authorization: `Bearer ${this.apiKey}` },
    });
    this.ws = ws;

    await new Promise<void>((resolve, reject) => {
      let settled = false;
      const settle = (fn: () => void): void => {
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        fn();
      };
      const timer = setTimeout(
        () =>
          settle(() =>
            reject(new PatterConnectionError('Speechmatics connect timeout')),
          ),
        CONNECT_TIMEOUT_MS,
      );
      ws.once('open', () => settle(resolve));
      ws.once('error', (err: Error) => settle(() => reject(err)));
      ws.once('unexpected-response', (_req: unknown, res: IncomingMessage) => {
        const status = res?.statusCode ?? 0;
        settle(() => {
          if (status === 401 || status === 403) {
            reject(
              new AuthenticationError(
                `Speechmatics rejected the API key (HTTP ${status}).`,
              ),
            );
            return;
          }
          if (status === 429) {
            reject(
              new RateLimitError('Speechmatics rate limit exceeded (HTTP 429).'),
            );
            return;
          }
          reject(
            new PatterConnectionError(
              `Speechmatics WebSocket upgrade failed (HTTP ${status}).`,
            ),
          );
        });
      });
    });

    ws.on('message', (raw) => this.handleMessage(raw.toString()));
    ws.on('close', () => this.handleClose());
    ws.on('error', (err: Error) => this.handleError(err));

    try {
      ws.send(JSON.stringify(this.buildStartRecognition()));
    } catch (err) {
      throw new PatterConnectionError(
        `Speechmatics StartRecognition send failed: ${String(err)}`,
      );
    }
    this.running = true;
  }

  /** Send a binary PCM16-LE audio chunk to Speechmatics for transcription. */
  sendAudio(audio: Buffer): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return;
    }
    if (audio.length === 0) {
      // Speechmatics treats empty frames as protocol errors — drop silently.
      return;
    }
    this.lastSeqNo += 1;
    try {
      this.ws.send(audio);
    } catch (err) {
      getLogger().error(`SpeechmaticsSTT sendAudio failed: ${String(err)}`);
    }
  }

  /** Register a transcript listener. */
  onTranscript(callback: TranscriptCallback): void {
    this.transcriptCallbacks.add(callback);
  }

  /** Remove a previously registered transcript listener. */
  offTranscript(callback: TranscriptCallback): void {
    this.transcriptCallbacks.delete(callback);
  }

  /** Register an error listener for socket / API failures. */
  onError(callback: ErrorCallback): void {
    this.errorCallbacks.add(callback);
  }

  /** Remove a previously registered error listener. */
  offError(callback: ErrorCallback): void {
    this.errorCallbacks.delete(callback);
  }

  private handleMessage(raw: string): void {
    let data: SpeechmaticsTranscriptMessage;
    try {
      data = JSON.parse(raw) as SpeechmaticsTranscriptMessage;
    } catch {
      return;
    }
    const event = data.message;
    if (!event) return;

    switch (event) {
      case SpeechmaticsServerMessage.RECOGNITION_STARTED:
      case SpeechmaticsServerMessage.AUDIO_ADDED:
      case SpeechmaticsServerMessage.END_OF_UTTERANCE:
      case SpeechmaticsServerMessage.END_OF_TRANSCRIPT:
      case SpeechmaticsServerMessage.INFO:
        // No transcript text — log only the lifecycle messages we care to
        // surface for debugging.
        return;
      case SpeechmaticsServerMessage.WARNING:
        getLogger().warn(`SpeechmaticsSTT warning: ${JSON.stringify(data)}`);
        return;
      case SpeechmaticsServerMessage.ERROR: {
        const message =
          data.reason ?? data.type ?? 'Speechmatics returned an Error frame';
        getLogger().error(`SpeechmaticsSTT error: ${message}`);
        this.emitError(new PatterConnectionError(`Speechmatics: ${message}`));
        return;
      }
      case SpeechmaticsServerMessage.ADD_PARTIAL_TRANSCRIPT:
      case SpeechmaticsServerMessage.ADD_TRANSCRIPT: {
        const isFinal = event === SpeechmaticsServerMessage.ADD_TRANSCRIPT;
        const transcript = this.toTranscript(data, isFinal);
        if (transcript !== null) this.emitTranscript(transcript);
        return;
      }
      default:
        return;
    }
  }

  /** Translate a Speechmatics transcript message into a Patter `Transcript`. */
  private toTranscript(
    message: SpeechmaticsTranscriptMessage,
    isFinal: boolean,
  ): Transcript | null {
    // Speechmatics RT supplies a ready-rendered `metadata.transcript` string
    // alongside the per-token `results[]`. Prefer the rendered transcript so
    // we don't have to re-implement word delimiter / punctuation rules.
    const rendered = (message.metadata?.transcript ?? '').trim();
    const results = message.results ?? [];

    let text = rendered;
    const confidences: number[] = [];
    for (const result of results) {
      const best = result.alternatives?.[0];
      if (!best) continue;
      const content = best.content;
      const confidence = best.confidence;
      if (!rendered && typeof content === 'string' && content.length > 0) {
        text = text ? `${text} ${content}` : content;
      }
      if (typeof confidence === 'number') {
        confidences.push(confidence);
      }
    }

    text = text.trim();
    if (!text) return null;

    const confidence =
      confidences.length > 0
        ? confidences.reduce((sum, c) => sum + c, 0) / confidences.length
        : 1.0;
    return { text, isFinal, confidence };
  }

  private emitTranscript(transcript: Transcript): void {
    for (const cb of this.transcriptCallbacks) {
      try {
        // The callback is async — the sync catch alone never saw its
        // rejection, which killed the process as an unhandled rejection.
        Promise.resolve(cb(transcript)).catch((err) =>
          getLogger().error(`SpeechmaticsSTT transcript callback failed: ${String(err)}`),
        );
      } catch (err) {
        getLogger().error(`SpeechmaticsSTT transcript callback threw: ${String(err)}`);
      }
    }
  }

  private emitError(err: Error): void {
    for (const cb of this.errorCallbacks) {
      try {
        cb(err);
      } catch (cbErr) {
        getLogger().error(`SpeechmaticsSTT error callback threw: ${String(cbErr)}`);
      }
    }
  }

  private handleError(err: Error): void {
    getLogger().error(`SpeechmaticsSTT WebSocket error: ${err.message}`);
    this.emitError(err);
  }

  private handleClose(): void {
    if (!this.running) return;
    this.running = false;
  }

  /** Send `EndOfStream` and close the WebSocket. Idempotent. */
  close(): void {
    this.running = false;
    const ws = this.ws;
    if (!ws) return;
    this.ws = null;

    const sendSafe = (payload: string): void => {
      if (ws.readyState === WebSocket.OPEN) {
        try {
          ws.send(payload);
        } catch {
          // ignore — the close path below tears down the socket anyway.
        }
      }
    };
    sendSafe(
      JSON.stringify({ message: 'EndOfStream', last_seq_no: this.lastSeqNo }),
    );

    try {
      ws.close();
    } catch {
      // ignore — best-effort close.
    }
  }
}

File: apps/patter-voice-sdk/libraries/python/getpatter/providers/speechmatics_stt.py
"""
Speechmatics Speech-to-Text adapter for the Patter SDK pipeline mode.

Uses the official ``speechmatics-voice[smart]`` SDK (imported lazily so the
dependency remains optional) to stream PCM audio to the Speechmatics real-time
API and yield :class:`~getpatter.providers.base.Transcript` events.

The audio-send / message-receive pipeline is exposed via a queue-fed
AsyncIterator, matching the pattern used by ``DeepgramSTT`` and
``WhisperSTT``. ``speechmatics.voice`` is imported lazily so consumers that
do not install the ``speechmatics`` extra can still import the rest of the
SDK.

Install with::

    pip install 'getpatter[speechmatics]'
"""

from __future__ import annotations

import asyncio
import logging
from enum import Enum, IntEnum, StrEnum
from typing import ClassVar, Any, AsyncIterator, Union

from getpatter.providers.base import STTProvider, Transcript

logger = logging.getLogger("getpatter")


SPEECHMATICS_INSTALL_HINT = (
    "The Speechmatics Voice SDK is required for SpeechmaticsSTT. "
    "Install the optional dependency with:\n"
    "    pip install 'getpatter[speechmatics]'\n"
    "or directly: pip install 'speechmatics-voice[smart]>=0.2.8'"
)


class TurnDetectionMode(str, Enum):
    """Endpoint / turn-detection handling mode.

    Mirrors the values accepted by ``speechmatics.voice.VoiceAgentConfigPreset``.
    See the Speechmatics docs for the semantic differences between modes.
    """

    EXTERNAL = "external"
    FIXED = "fixed"
    ADAPTIVE = "adaptive"
    SMART_TURN = "smart_turn"


class SpeechmaticsSampleRate(IntEnum):
    """Common PCM sample rates for Speechmatics streaming input."""

    HZ_8000 = 8000
    HZ_16000 = 16000
    HZ_44100 = 44100


class SpeechmaticsAudioEncoding(StrEnum):
    """Audio encodings accepted by Speechmatics's real-time API."""

    PCM_S16LE = "pcm_s16le"


class SpeechmaticsOperatingPoint(StrEnum):
    """Speechmatics operating points (accuracy vs latency trade-off)."""

    ENHANCED = "enhanced"
    STANDARD = "standard"


def _require_voice_sdk() -> Any:
    """Import and return the ``speechmatics.voice`` module.

    Raises a ``RuntimeError`` with installation instructions if the SDK is
    not available, rather than leaking an ``ImportError`` at import time.
    """
    try:
        import speechmatics.voice as voice_sdk  # type: ignore[import-not-found]
    except ImportError as exc:  # pragma: no cover - trivial guard
        raise RuntimeError(SPEECHMATICS_INSTALL_HINT) from exc
    return voice_sdk


class SpeechmaticsSTT(STTProvider):
    """Speechmatics real-time STT adapter.

    Connects via the official ``speechmatics-voice`` SDK and emits
    :class:`Transcript` objects for partial and final segments.

    Args:
        api_key: Speechmatics API key (or ``SPEECHMATICS_API_KEY`` env var).
        base_url: Optional override for the Speechmatics realtime endpoint
            (``SPEECHMATICS_RT_URL`` env var, falls back to the SDK default).
        language: BCP-47 language code (default ``"en"``).
        turn_detection_mode: How Speechmatics detects end of turn. Use
            ``TurnDetectionMode.EXTERNAL`` with an external VAD.
        sample_rate: PCM sample rate (Hz). Defaults to 16 kHz.
        enable_diarization: Attach speaker IDs to transcripts.
        max_delay: Max latency in seconds before the engine emits finals.
        end_of_utterance_silence_trigger: Silence (s) that triggers EOU.
        end_of_utterance_max_delay: Max EOU delay (s); must be greater
            than ``end_of_utterance_silence_trigger``.
        include_partials: Include partial words in interim output.
        additional_vocab: List of ``AdditionalVocabEntry`` from the SDK.
        operating_point: SDK ``OperatingPoint`` enum (accuracy vs latency).
        domain: Optional Speechmatics domain (e.g. ``"finance"``).
        output_locale: Optional output locale (e.g. ``"en-GB"``).
    """

    #: Stable pricing/dashboard key — read by stream-handler/metrics.
    provider_key: ClassVar[str] = "speechmatics"

    # Sentinel used to shut down the receive loop.
    _STOP = object()

    def __init__(
        self,
        api_key: str,
        *,
        base_url: str | None = None,
        language: str = "en",
        turn_detection_mode: TurnDetectionMode = TurnDetectionMode.ADAPTIVE,
        sample_rate: Union[
            SpeechmaticsSampleRate, int
        ] = SpeechmaticsSampleRate.HZ_16000,
        enable_diarization: bool = False,
        max_delay: float | None = None,
        end_of_utterance_silence_trigger: float | None = None,
        end_of_utterance_max_delay: float | None = None,
        include_partials: bool = True,
        additional_vocab: list[Any] | None = None,
        operating_point: Any | None = None,
        domain: str | None = None,
        output_locale: str | None = None,
    ) -> None:
        if not api_key:
            raise ValueError("Speechmatics api_key is required")

        # Validate ranges per the Speechmatics Voice SDK contract.
        if end_of_utterance_silence_trigger is not None and not (
            0 < end_of_utterance_silence_trigger < 2
        ):
            raise ValueError("end_of_utterance_silence_trigger must be between 0 and 2")
        if (
            end_of_utterance_max_delay is not None
            and end_of_utterance_silence_trigger is not None
            and end_of_utterance_max_delay <= end_of_utterance_silence_trigger
        ):
            raise ValueError(
                "end_of_utterance_max_delay must be greater than "
                "end_of_utterance_silence_trigger"
            )
        if max_delay is not None and not (0.7 <= max_delay <= 4.0):
            raise ValueError("max_delay must be between 0.7 and 4.0")

        # Eagerly validate SDK availability so failures surface at init.
        self._voice = _require_voice_sdk()

        self.api_key = api_key
        self.base_url = base_url
        self.language = language
        self.turn_detection_mode = turn_detection_mode
        self.sample_rate = sample_rate
        self.enable_diarization = enable_diarization
        self.max_delay = max_delay
        self.end_of_utterance_silence_trigger = end_of_utterance_silence_trigger
        self.end_of_utterance_max_delay = end_of_utterance_max_delay
        self.include_partials = include_partials
        self.additional_vocab = additional_vocab or []
        self.operating_point = operating_point
        self.domain = domain
        self.output_locale = output_locale

        self._client: Any | None = None
        self._queue: asyncio.Queue[Any] = asyncio.Queue()
        # Speechmatics always streams pcm_s16le (see ``_build_config``);
        # expose encoding for observability cost computation.
        self.encoding: str = "linear16"
        self._audio_bytes_sent: int = 0

    def _record_transcript_cost(self) -> None:
        """Emit ``patter.cost.stt_seconds`` for buffered audio."""
        try:
            from getpatter.observability.attributes import record_patter_attrs

            bytes_per_sample = 1 if self.encoding == "mulaw" else 2
            seconds = self._audio_bytes_sent / float(
                self.sample_rate * bytes_per_sample
            )
            record_patter_attrs(
                {
                    "patter.cost.stt_seconds": seconds,
                    "patter.stt.provider": "speechmatics",
                }
            )
            self._audio_bytes_sent = 0
        except Exception:  # pragma: no cover — defense in depth
            logger.debug("_record_transcript_cost failed", exc_info=True)

    def __repr__(self) -> str:
        return (
            f"SpeechmaticsSTT(language={self.language!r}, "
            f"turn_detection={self.turn_detection_mode.value!r}, "
            f"sample_rate={self.sample_rate})"
        )

    # ------------------------------------------------------------------
    # Connection lifecycle
    # ------------------------------------------------------------------

    def _build_config(self) -> Any:
        """Build a ``VoiceAgentConfig`` from the adapter parameters."""
        voice = self._voice
        config = voice.VoiceAgentConfigPreset.load(self.turn_detection_mode.value)
        config.sample_rate = self.sample_rate
        config.audio_encoding = voice.AudioEncoding.PCM_S16LE
        config.language = self.language
        if self.domain is not None:
            config.domain = self.domain
        if self.output_locale is not None:
            config.output_locale = self.output_locale
        if self.additional_vocab:
            config.additional_vocab = self.additional_vocab
        if self.operating_point is not None:
            config.operating_point = self.operating_point
        if self.max_delay is not None:
            config.max_delay = self.max_delay
        if self.end_of_utterance_silence_trigger is not None:
            config.end_of_utterance_silence_trigger = (
                self.end_of_utterance_silence_trigger
            )
        if self.end_of_utterance_max_delay is not None:
            config.end_of_utterance_max_delay = self.end_of_utterance_max_delay
        config.enable_diarization = self.enable_diarization
        config.include_partials = self.include_partials
        return config

    def _register_handlers(self, client: Any) -> None:
        """Wire Speechmatics server events to our internal queue."""
        voice = self._voice

        def _enqueue(message: dict[str, Any]) -> None:
            self._queue.put_nowait(message)

        handlers = [
            voice.AgentServerMessageType.ADD_PARTIAL_SEGMENT,
            voice.AgentServerMessageType.ADD_SEGMENT,
            voice.AgentServerMessageType.END_OF_TURN,
            voice.AgentServerMessageType.ERROR,
            voice.AgentServerMessageType.WARNING,
        ]
        for event in handlers:
            client.on(event, _enqueue)

    async def connect(self) -> None:
        """Create the underlying ``VoiceAgentClient`` and connect."""
        if self._client is not None:
            return

        voice = self._voice
        kwargs: dict[str, Any] = {
            "api_key": self.api_key,
            "config": self._build_config(),
            "app": "patter-sdk",
        }
        if self.base_url:
            kwargs["url"] = self.base_url

        self._client = voice.VoiceAgentClient(**kwargs)
        self._register_handlers(self._client)
        # Drain a stale _STOP sentinel from a previous close() so a
        # sequential second call on the same instance doesn't terminate its
        # transcript loop on the first get().
        while not self._queue.empty():
            try:
                self._queue.get_nowait()
            except asyncio.QueueEmpty:  # pragma: no cover - defensive
                break
        await self._client.connect()

    async def send_audio(self, audio_chunk: bytes) -> None:
        """Send a PCM audio chunk (signed 16-bit little-endian)."""
        if self._client is None:
            raise RuntimeError(
                "SpeechmaticsSTT is not connected. Call connect() first."
            )
        if not audio_chunk:
            return
        self._audio_bytes_sent += len(audio_chunk)
        await self._client.send_audio(audio_chunk)

    # ------------------------------------------------------------------
    # Transcript stream
    # ------------------------------------------------------------------

    def _message_to_transcript(self, message: dict[str, Any]) -> Transcript | None:
        """Translate a Speechmatics server message into a Patter Transcript."""
        voice = self._voice
        event = message.get("message")

        if event == voice.AgentServerMessageType.ADD_PARTIAL_SEGMENT:
            is_final = False
        elif event == voice.AgentServerMessageType.ADD_SEGMENT:
            is_final = True
        elif event == voice.AgentServerMessageType.END_OF_TURN:
            # End-of-turn carries no text; Patter signals finality via the
            # is_final flag on transcript events — ignore here.
            return None
        elif event in (
            voice.AgentServerMessageType.ERROR,
            voice.AgentServerMessageType.WARNING,
        ):
            logger.warning("SpeechmaticsSTT %s: %s", event, message)
            return None
        else:
            return None

        segments = message.get("segments") or []
        texts: list[str] = []
        confidences: list[float] = []
        for segment in segments:
            text = segment.get("text")
            if text:
                texts.append(text)
            confidence = segment.get("confidence")
            if isinstance(confidence, (int, float)):
                confidences.append(float(confidence))

        joined = " ".join(t for t in texts if t).strip()
        if not joined:
            return None

        confidence = sum(confidences) / len(confidences) if confidences else 1.0
        return Transcript(text=joined, is_final=is_final, confidence=confidence)

    async def receive_transcripts(self) -> AsyncIterator[Transcript]:
        """Yield :class:`Transcript` objects from Speechmatics events."""
        if self._client is None:
            raise RuntimeError(
                "SpeechmaticsSTT is not connected. Call connect() first."
            )

        while True:
            message = await self._queue.get()
            if message is self._STOP:
                break
            try:
                transcript = self._message_to_transcript(message)
            except Exception as exc:  # noqa: BLE001
                logger.exception("SpeechmaticsSTT handler error: %s", exc)
                continue
            if transcript is not None:
                if transcript.is_final:
                    self._record_transcript_cost()
                yield transcript

    # ------------------------------------------------------------------
    # Cleanup
    # ------------------------------------------------------------------

    async def close(self) -> None:
        """Disconnect the Speechmatics client and stop the receive loop."""
        if self._client is not None:
            try:
                await self._client.disconnect()
            except Exception as exc:  # noqa: BLE001
                logger.debug("SpeechmaticsSTT disconnect error: %s", exc)
            self._client = None

        # Unblock any outstanding receive_transcripts() call.
        try:
            self._queue.put_nowait(self._STOP)
        except Exception:  # noqa: BLE001
            pass
